File: packages/SyncCore/Sources/SyncCore/FrameCodec.swift
import Foundation

public enum FrameKind: UInt8, Codable, Equatable, Sendable {
    case session = 1
    case offer = 2
    case decision = 3
    case chunk = 4
    case result = 5
}

public enum FrameCodecError: Error, Equatable, Sendable {
    case invalidMagic
    case unsupportedVersion(UInt16)
    case invalidKind(UInt8)
    case invalidHeaderLength
    case payloadTooLarge
    case truncatedFrame
    case trailingBytes
    case controlKindMismatch
}

public struct FrameHeader: Equatable, Sendable {
    public static let byteCount = 40

    public let version: UInt16
    public let kind: FrameKind
    public let requestID: UUID
    public let offset: UInt64
    public let payloadLength: UInt64
}

public struct SyncFrame: Equatable, Sendable {
    public let kind: FrameKind
    public let requestID: UUID
    public let offset: UInt64
    public let payload: Data

    public init(kind: FrameKind, requestID: UUID, offset: UInt64, payload: Data) throws {
        let maximum = kind == .chunk ? SyncConstants.chunkSize : SyncConstants.maximumControlPayload
        guard payload.count <= maximum else {
            throw FrameCodecError.payloadTooLarge
        }
        self.kind = kind
        self.requestID = requestID
        self.offset = offset
        self.payload = payload
    }

    public static func control(_ message: SyncControlMessage, requestID: UUID) throws -> SyncFrame {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let payload = try encoder.encode(message)
        return try SyncFrame(kind: message.frameKind, requestID: requestID, offset: 0, payload: payload)
    }

    public func controlMessage() throws -> SyncControlMessage {
        guard kind != .chunk else {
            throw FrameCodecError.controlKindMismatch
        }
        let message = try JSONDecoder().decode(SyncControlMessage.self, from: payload)
        guard message.frameKind == kind else {
            throw FrameCodecError.controlKindMismatch
        }
        return message
    }
}

public enum FrameCodec {
    private static let magic: [UInt8] = [0x49, 0x50, 0x53, 0x31]

    public static func encode(_ frame: SyncFrame) throws -> Data {
        let validated = try SyncFrame(
            kind: frame.kind,
            requestID: frame.requestID,
            offset: frame.offset,
            payload: frame.payload
        )
        var data = Data()
        data.reserveCapacity(FrameHeader.byteCount + validated.payload.count)
        data.append(contentsOf: magic)
        appendUInt16(SyncConstants.protocolVersion, to: &data)
        data.append(validated.kind.rawValue)
        data.append(0)

        var uuid = validated.requestID.uuid
        withUnsafeBytes(of: &uuid) { data.append(contentsOf: $0) }

        appendUInt64(validated.offset, to: &data)
        appendUInt64(UInt64(validated.payload.count), to: &data)
        data.append(validated.payload)
        return data
    }

    public static func decodeHeader(_ data: Data) throws -> FrameHeader {
        guard data.count == FrameHeader.byteCount else {
            throw FrameCodecError.invalidHeaderLength
        }
        guard Array(data[0..<4]) == magic else {
            throw FrameCodecError.invalidMagic
        }

        let version = readUInt16(data, at: 4)
        guard version == SyncConstants.protocolVersion else {
            throw FrameCodecError.unsupportedVersion(version)
        }
        let rawKind = data[6]
        guard let kind = FrameKind(rawValue: rawKind) else {
            throw FrameCodecError.invalidKind(rawKind)
        }

        let bytes = Array(data[8..<24])
        let requestID = UUID(uuid: (
            bytes[0], bytes[1], bytes[2], bytes[3],
            bytes[4], bytes[5], bytes[6], bytes[7],
            bytes[8], bytes[9], bytes[10], bytes[11],
            bytes[12], bytes[13], bytes[14], bytes[15]
        ))
        let offset = readUInt64(data, at: 24)
        let payloadLength = readUInt64(data, at: 32)
        let maximum = kind == .chunk ? SyncConstants.chunkSize : SyncConstants.maximumControlPayload
        guard payloadLength <= UInt64(maximum) else {
            throw FrameCodecError.payloadTooLarge
        }

        return FrameHeader(
            version: version,
            kind: kind,
            requestID: requestID,
            offset: offset,
            payloadLength: payloadLength
        )
    }

    public static func decodeCompleteFrame(_ data: Data) throws -> SyncFrame {
        guard data.count >= FrameHeader.byteCount else {
            throw FrameCodecError.truncatedFrame
        }
        let headerData = data.prefix(FrameHeader.byteCount)
        let header = try decodeHeader(Data(headerData))
        let expectedCount = FrameHeader.byteCount + Int(header.payloadLength)
        guard data.count >= expectedCount else {
            throw FrameCodecError.truncatedFrame
        }
        guard data.count == expectedCount else {
            throw FrameCodecError.trailingBytes
        }
        return try SyncFrame(
            kind: header.kind,
            requestID: header.requestID,
            offset: header.offset,
            payload: Data(data.dropFirst(FrameHeader.byteCount))
        )
    }

    private static func appendUInt16(_ value: UInt16, to data: inout Data) {
        data.append(UInt8((value >> 8) & 0xff))
        data.append(UInt8(value & 0xff))
    }

    private static func appendUInt64(_ value: UInt64, to data: inout Data) {
        for shift in stride(from: 56, through: 0, by: -8) {
            data.append(UInt8((value >> UInt64(shift)) & 0xff))
        }
    }

    private static func readUInt16(_ data: Data, at offset: Int) -> UInt16 {
        (UInt16(data[offset]) << 8) | UInt16(data[offset + 1])
    }

    private static func readUInt64(_ data: Data, at offset: Int) -> UInt64 {
        var value: UInt64 = 0
        for index in offset..<(offset + 8) {
            value = (value << 8) | UInt64(data[index])
        }
        return value
    }
}

File: packages/SyncCore.Windows/src/protocol/frame-codec.ts
// FrameCodec — 40-byte big-endian header + payload, mirroring
// `packages/SyncCore/Sources/SyncCore/FrameCodec.swift`.
//
// Header layout (offset, bytes, field):
//   0   4   magic           ASCII "IPS1" = 0x49 0x50 0x53 0x31
//   4   2   version         u16 BE, must equal SyncConstants.protocolVersion
//   6   1   kind            u8  (FrameKind rawValue)
//   7   1   reserved        u8  (must be 0)
//   8  16   requestID       raw UUID bytes
//  24   8   offset          u64 BE
//  32   8   payloadLength   u64 BE
//  40   N   payload

import { randomUUID } from 'node:crypto';
import { SyncConstants } from './constants.js';

export const FRAME_MAGIC = Uint8Array.from([0x49, 0x50, 0x53, 0x31]); // "IPS1"
export const FRAME_HEADER_LENGTH = 40;

export enum FrameKind {
  session = 1,
  offer = 2,
  decision = 3,
  chunk = 4,
  result = 5,
}

export interface FrameHeader {
  version: number;
  kind: FrameKind;
  reserved: number;
  requestID: Uint8Array; // 16 bytes
  offset: bigint;
  payloadLength: bigint;
}

export class FrameCodecError extends Error {
  constructor(
    message: string,
    public readonly code:
      | 'invalidMagic'
      | 'unsupportedVersion'
      | 'payloadTooLarge'
      | 'truncatedFrame'
      | 'invalidHeaderLength'
      | 'invalidKind',
  ) {
    super(message);
    this.name = 'FrameCodecError';
  }
}

export class FrameCodec {
  /** Encode a header + payload into a single `Uint8Array`. */
  static encode(header: FrameHeader, payload: Uint8Array): Uint8Array {
    if (payload.length > Number.MAX_SAFE_INTEGER) {
      throw new RangeError('payload too large to encode');
    }
    const payloadLength = BigInt(payload.length);
    if (header.kind === FrameKind.chunk) {
      if (payloadLength > BigInt(SyncConstants.chunkSize)) {
        throw new FrameCodecError(
          `chunk payload ${payloadLength} exceeds chunkSize ${SyncConstants.chunkSize}`,
          'payloadTooLarge',
        );
      }
    } else if (payloadLength > BigInt(SyncConstants.maximumControlPayload)) {
      throw new FrameCodecError(
        `control payload ${payloadLength} exceeds maximumControlPayload ${SyncConstants.maximumControlPayload}`,
        'payloadTooLarge',
      );
    }
    if (header.requestID.length !== 16) {
      throw new FrameCodecError('requestID must be 16 bytes', 'invalidHeaderLength');
    }
    const out = new Uint8Array(FRAME_HEADER_LENGTH + payload.length);
    out.set(FRAME_MAGIC, 0);
    const view = new DataView(out.buffer);
    view.setUint16(4, header.version, false);
    view.setUint8(6, header.kind);
    view.setUint8(7, header.reserved);
    out.set(header.requestID, 8);
    view.setBigUint64(24, header.offset, false);
    view.setBigUint64(32, payloadLength, false);
    out.set(payload, FRAME_HEADER_LENGTH);
    return out;
  }

  /** Decode a header from a 40-byte buffer. */
  static decodeHeader(buffer: Uint8Array): FrameHeader {
    if (buffer.length < FRAME_HEADER_LENGTH) {
      throw new FrameCodecError(
        `truncated header: ${buffer.length} bytes`,
        'truncatedFrame',
      );
    }
    for (let i = 0; i < 4; i++) {
      if (buffer[i] !== FRAME_MAGIC[i]) {
        throw new FrameCodecError('invalid magic', 'invalidMagic');
      }
    }
    const view = new DataView(buffer.buffer, buffer.byteOffset, FRAME_HEADER_LENGTH);
    const version = view.getUint16(4, false);
    if (version !== SyncConstants.protocolVersion) {
      throw new FrameCodecError(
        `unsupported protocol version ${version}`,
        'unsupportedVersion',
      );
    }
    const kind = view.getUint8(6);
    if (!(kind in FrameKind)) {
      throw new FrameCodecError(`invalid frame kind ${kind}`, 'invalidKind');
    }
    const reserved = view.getUint8(7);
    const requestID = buffer.slice(8, 24);
    const offset = view.getBigUint64(24, false);
    const payloadLength = view.getBigUint64(32, false);
    return {
      version,
      kind: kind as FrameKind,
      reserved,
      requestID,
      offset,
      payloadLength,
    };
  }

  static newRequestID(): Uint8Array {
    // RFC 4122 random UUID; raw 16 bytes match Swift's `UUID().uuid`.
    const hex = randomUUID().replace(/-/g, '');
    const out = new Uint8Array(16);
    for (let i = 0; i < 16; i++) {
      out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return out;
  }
}
